pkg/ccl/changefeedccl/testfeed_test.go

```diff
@@ -1041,52 +1058,141 @@ func extractKeyFromJSONValue(isBare bool, wrapped []byte) (key []byte, value []b
+	}
+	defer f.Close()
+
+	fr, err := goparquet.NewFileReader(f)
+	if err != nil {
+		return err
+	}
+
+	keyCols, ok := fr.MetaData()["pkeys"]
+	if !ok {
+		return err
+	}
+	keyColsSet := make(map[string]struct{})
+	for _, key := range strings.Split(keyCols, ",") {
+		keyColsSet[key] = struct{}{}
+	}
+
 	for {
-		if len(c.rows) > 0 {
-			e := c.rows[0]
-			c.rows = c.rows[1:]
-			m := &cdctest.TestFeedMessage{
-				Topic:    e.topic,
-				Value:    e.value,
-				Resolved: e.payload,
-			}
 
-			// The other TestFeed impls check both key and value here, but cloudFeeds
-			// don't have keys.
-			if len(m.Value) > 0 {
-				details, err := c.Details()
-				if err != nil {
-					return nil, err
-				}
+		row, err := fr.NextRow()
+		if err == io.EOF {
+			break
+		}
 
-				switch v := changefeedbase.FormatType(details.Opts[changefeedbase.OptFormat]); v {
-				case ``, changefeedbase.OptFormatJSON:
-					// Cloud storage sinks default the `WITH key_in_value` option so that
-					// the key is recoverable. Extract it out of the value (also removing it
-					// so the output matches the other sinks). Note that this assumes the
-					// format is json, this will have to be fixed once we add format=avro
-					// support to cloud storage.
-					//
-					// TODO(dan): Leave the key in the value if the TestFeed user
-					// specifically requested it.
-					if m.Key, m.Value, err = extractKeyFromJSONValue(c.isBare, m.Value); err != nil {
-						return nil, err
-					}
-					if isNew := c.markSeen(m); !isNew {
-						continue
+		if err != nil {
+			return err
+		}
+
+		// Holds column to its value mapping for a row and the entire thing will be
+		// JSON encoded later.
+		value := make(map[string]interface{})
+
+		// Holds the list of values of primary keys.
+		key := make([]interface{}, 0)
+
+		for k, v := range row {
+
+			switch vv := v.(type) {
+			case []byte:
+				// Currently, for encoding from CRDB data type to Parquet data type, for
+				// any complex structure (that is, other than ints and floats), it is
+				// always a byte array which is the string representation of that datum
+				// (except for arrays, see below). Therefore, if the parquet reader
+				// decodes a column value as a Go native byte array, then we can be sure
+				// that it is the equivalent string representation of the CRDB datum.
+				// Hence, we can use this value to construct the final JSON object which
+				// will be used by assertPayload to compare actual and expected JSON
+				// objects.
+
+				// Ideally there should be no need to convert byte array to string but
+				// JSON encoder will encode byte arrays as base 64 encoded strings.
+				// Hence, we need to convert to string to tell Marshal to decode it as a
+				// string. For every other Go native type, we can use the type as is and
+				// json.Marhsal will work correctly.
+				value[k] = string(vv)
+			case map[string]interface{}:
+				// This is CRDB ARRAY data type (only data type for which we use parquet
```

Comment:
it's strange to see array, when the case is for map[string]interface -- which looks like an object.